### PR TITLE
Add admin filters

### PR DESCRIPTION
**Pull request by core23**
## Subject

Adds reuseable filters that can respect the admins classification context.

I am targeting this branch, because this feature is BC.

## Changelog

```markdown
### Added
- Added `CategoryFilter` for admin lists
- Added `CollectionFilter` for admin lists
```

### CODE REVIEW

File: src/Admin/Filter/CategoryFilter.php [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+    public function getFieldOptions()
+    {
+        return $this->getOption('choices', [
+            'choices' => $this->getChoices(),
+            'choice_translation_domain' => false,
+        ]);
+    }
+
+    public function getRenderSettings()
```
Comment (greg0ire) [inline src/Admin/Filter/CategoryFilter.php]: Please add return type hints, and for the other class too

File: src/Admin/Filter/CategoryFilter.php [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+            $queryBuilder
+                ->andWhere(sprintf('%s.%s = :category', $alias, $field))
+                ->setParameter('category', $data['value'])
+            ;
+        }
+
+        $this->active = null !== $data['value'];
+    }
+
+    public function getDefaultOptions(): array
```
Comment (OskarStark) [inline src/Admin/Filter/CategoryFilter.php]:
```suggestion
    /**
    * @return array<string, null>
    */
    public function getDefaultOptions(): array
```

Shall we add those for further static code analysis support?

File: src/Admin/Filter/CollectionFilter.php [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+    public function __construct(CollectionManagerInterface $collectionManager)
+    {
+        $this->collectionManager = $collectionManager;
+    }
+
+    /**
+     * @param string            $alias
+     * @param string            $field
+     * @param array|string|null $data
```
Comment (VincentLanglet) [inline src/Admin/Filter/CollectionFilter.php:38]: I see everywhere in these bundle checks like is_array for the $data value.

But according the interface, it's always an array
image

File: src/Admin/Filter/CategoryFilter.php [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        $this->categoryManager = $categoryManager;
+    }
+
+    public function filter(ProxyQueryInterface $queryBuilder, $alias, $field, $data): void
+    {
+        if (null === $data || !\is_array($data) || !\array_key_exists('value', $data)) {
+            return;
+        }
+
+        if ($data['value']) {
```
Comment (phansys) [inline src/Admin/Filter/CategoryFilter.php:41]: Could not `$data['value']` hold `null`, `0`, `false` or `''`?

File: tests/Admin/Filter/QueryBuilder.php [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    public function setParameter($key, $value, $type = null)
+    {
+        $this->parameters[$key] = $value;
+    }
+
+    public function andWhere()
+    {
+        $query = \func_get_args();
+
+        $this->query[] = (string) $query;
```
Comment (phansys) [inline tests/Admin/Filter/QueryBuilder.php:42]: Isn't this lead to an "array to string conversion" error?

File: src/Admin/Filter/CategoryFilter.php [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        foreach ($categories as $category) {
+            $choices[sprintf('%s (%s)', $category->getName(), $category->getContext()->getId())] = $category->getId();
+
+            $this->childWalker($category, $choices);
+        }
+
+        return $choices;
+    }
+
+    private function childWalker(CategoryInterface $category, array &$choices, int $level = 2): void
```
Comment (greg0ire) [inline src/Admin/Filter/CategoryFilter.php:113]: naming a method as a noun is unexpected, it's not even an accessor. How about `visitChild` instead

File: tests/Admin/Filter/CollectionFilterTest.php [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+class CollectionFilterTest extends TestCase
+{
+    /**
+     * @var MockObject&CollectionManagerInterface
+     */
+    private $collectionManager;
+
+    protected function setUp(): void
+    {
+        $this->collectionManager = $this->createMock(CollectionManagerInterface::class);
```
Comment (greg0ire) [inline tests/Admin/Filter/CollectionFilterTest.php:31]:
```suggestion
        $this->collectionManager = $this->createStub(CollectionManagerInterface::class);
```

File: tests/Admin/Filter/QueryBuilder.php [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\ClassificationBundle\Tests\Admin\Filter;
+
+use Doctrine\ORM\Query\Expr\Andx;
+use Doctrine\ORM\Query\Expr\Orx;
+
+final class QueryBuilder extends \Doctrine\ORM\QueryBuilder
```
Comment (greg0ire) [inline tests/Admin/Filter/QueryBuilder.php:19]: Why not use return type hints in this class?

File: tests/Admin/Filter/QueryBuilder.php (hunk 5 above)
Comment (core23) [reply tests/Admin/Filter/QueryBuilder.php:42]: It's basically a copy & paste of https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/3.x/tests/Filter/QueryBuilder.php

Comment (phansys) [reply tests/Admin/Filter/QueryBuilder.php:42]: I don't see where we are casting the result of `func_get_args()` as string at sonata-project/doctrine-orm-admin-bundle.
Just for the records, here is the example: https://3v4l.org/6UZbV.

**Comment by core23** [comment]
Please review @sonata-project/contributors

File: src/Admin/Filter/CollectionFilter.php [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+        if (null === $context) {
+            $collections = $this->collectionManager->findAll();
+        } else {
+            $collections = $this->collectionManager->getByContext($context);
+        }
+
+        $choices = [];
+
+        /** @var CollectionInterface $collection */
```
Comment (greg0ire) [inline src/Admin/Filter/CollectionFilter.php:104]: Why is this needed?
